accepted/wasm-gc.md

```diff
@@ -0,0 +1,1362 @@
+    pub fn new(engine: &Engine, element_type: FieldType) -> ArrayType { /* ... */ }
+
+    pub fn element_type(&self, engine: &Engine) -> FieldType { /* ... */ }
+}
+
+pub struct FunctionType(u32);
+
+impl FunctionType {
+    pub fn new(
+        engine: &Engine,
+        params: impl IntoIterator<Item = ValType>,
+        results: impl IntoIterator<Item = ValType>,
+    ) -> FunctionType {
+        // ...
+    }
+
+    pub fn params<'engine>(&self, engine: &'engine Engine) -> &'engine [ValType] { /* ... */ }
+
+    pub fn results<'engine>(&self, engine: &'engine Engine) -> &'engine [ValType] { /* ... */ }
+}
+
+pub struct StructType(u32);
+
+impl StructType {
+    pub fn new(
+        engine: &Engine,
+        fields: impl IntoIterator<Item = FieldType>,
+    ) -> StructType {
+        // ...
+    }
+
+    pub fn fields<'engine>(&self, engine: &'engine Engine) -> &'engine [FieldType] { /* ... */ }
+}
+
+// (These types follow directly from the Wasm GC proposal, but are included for
+// completeness since they are referenced above.)
+
+pub enum PackedType {
+    I8,
+    I16,
+}
+
+pub enum StorageType {
+    Packed(PackedType),
+    Unpacked(ValType),
+}
+
+pub struct FieldType {
+    pub mutable: bool,
+    pub storage: StorageType,
+}
+```
+
+Note that this proposed `FunctionType` will replace the existing
+`wasmtime::FuncType` and will be an annoying (but mechanical) breaking API
+change.[^breaking-api]
+
+[^breaking-api]: In general, we can't avoid breaking changes to the public
+    `wasmtime` crate's API while exposing GC references. At minimum we need to
+    add new variants to `enum`s.
+
+### Values
+
+As for exposing reference values in the public API, we have three major goals,
+which are somewhat in conflict:
+
+1. Ergonomics
+2. Performance
+3. Safety
+
+(3) is a hard requirement, at least for now. I think we need *something* working
+that is 100% safe, and if we need to add unsafe APIs on the side to release
+specific pressures (similar to `ValRaw`) then we can do that, but it would be a
+colossal failure for us to lose our API's safe-by-default property and force
+unsafety on all Wasmtime embedders.
+
+Additionally, there are three disjoint type hierarchies in the Wasm GC proposal:
+
+1. Internal data types
+2. Function types
+3. Extern types
+
+These have the following super-->sub type relationships:
+
+![](./wasm-gc-type-hierarchy.png)
+
+This RFC proposes creating a Rust type that mirrors each of the types that are
+in every Wasm module, but not a Rust type for Wasm types that reference a type
+that is defined in the module's type section.
+
+With those goals in mind, here is the proposed API for exposing reference values
+in the public API. Similar to `wasmtime::ValType::Ref`, we will add a new `Ref`
+variant of `wasmtime::Val`, which itself is an `enum` over the three different
+Wasm type hierarchies:
+
+```rust
+pub enum Val {
+    // <existing variants>
+
+    Ref(Ref),
+}
+
+pub enum Ref {
+    Internal(AnyRef),
```

Comment:
Right now all `{Any,Eq,Struct,Array}Ref`s are nullable on the Rust side.

That is, as written we don't represent null with `None` and non-null with `Some(FooRef)` and non-nullable foo refs with `FooRef`.

This is a good point to consider though. Should we switch to that representation? It is actually what we do with `funcref` and `externref` right now in the public API, but it does mean that you can't get precise type information for a null reference. I guess that is maybe not a super useful property to have though.

@alexcrichton do you happen to have thoughts on this?

Reply:
I think the useful ability is to avoid null checks when you statically know that a value going into an operation is non-nullable anyway. I think all instructions accept nullable values, so you can probably get away with not expressing nullability in your value representation, but why leave performance on the table?

Reply:
> but it does mean that you can't get precise type information for a null reference

I'm not sure what you mean by this @fitzgen? Otherwise I'd naively say that `Option<T>` would be the better representation than internally storing the fact that it's null. I may not understand what else is in play here, but I would assume that the precise type information for the reference comes from the `T`?